### PR TITLE
Csr feat/implement serializer layer

### PR DESCRIPTION
**Pull request by aggarw13**
*Issue #, if available:*
[TS-12368](https://issues.amazon.com/issues/TS-12368)

*Description of changes:*
* Implement the serializer function for the request payload of the new Certificate-Signing Request feature in Fleet Provisioning library.
* Add unit tests for the serializer.

CSR feature specification document: https://w.amazon.com/bin/view/AWS/Mobile/IoT/Foundry/Internal/FoundryCSRSupport/#HAPIOverview

This PR is preceded by another PR on the public repository: https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/826

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: libraries/aws/provisioning/src/aws_iot_provisioning_api.c [hunk 1]
```diff
@@ -617,10 +617,10 @@ AwsIotProvisioningError_t AwsIotProvisioning_CreateCertificateFromCsr( IotMqttCo
         /* Serialization of request payload occurs in a 2-step process, one for calculation of buffer size, and then,
```
Comment (gkwicker) [inline libraries/aws/provisioning/src/aws_iot_provisioning_api.c]: I'd add to this comment that the function returns the size even when the buffer address is `NULL`.

File: libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c [hunk 2]
```diff
@@ -346,6 +346,11 @@ AwsIotProvisioningError_t _AwsIotProvisioning_SerializeCreateKeysAndCertificateR
                                                                *pSerializationBuffer,
                                                                *pBufferSize );
 
+    if( status != AWS_IOT_PROVISIONING_SUCCESS )
```
Comment (gkwicker) [inline libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c]: Make sure we have a task to eliminate these `goto`s.

File: libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c [hunk 3]
```diff
@@ -355,17 +360,111 @@ AwsIotProvisioningError_t _AwsIotProvisioning_SerializeCreateKeysAndCertificateR
 
 /*------------------------------------------------------------------*/
 
-bool _AwsIotProvisioning_SerializeCreateCertificateFromCsrRequestPayload( const char * pCertificateSigningRequest,
-                                                                          size_t csrLength,
-                                                                          uint8_t * pSerializationBuffer,
-                                                                          size_t * pBufferSize )
+bool _AwsIotProvisioning_SerializeCreateCertFromCsrRequestPayload( const char * pCertificateSigningRequest,
```
Comment (gkwicker) [inline libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c]: This function is over-complicated because it has two different jobs.  It has the job described by the name of the function, which is to serialize a create certificate request from the CSR payload, but it has a second job which is to determine how much buffer space is required to do the first.  Because we have these two jobs, we end up with 3 different places where we're checking to see which job we're doing: line 374, 389, and line 454.

To fix this, create a `static` helper function for job 2 (calculating the size required).  That way, you'll only need to test `pSerializationBuffer == NULL` once at line 374 and this function will be a lot simpler and easier to read.

File: libraries/aws/provisioning/test/aws_iot_tests_provisioning.c [hunk 4]
```diff
@@ -44,13 +44,13 @@ void RunProvisioningTests( bool disableNetworkTests,
     /* Silence warnings about unused parameters. */
     ( void ) disableLongTests;
 
-    RUN_TEST_GROUP( Provisioning_Unit_API );
+    /* RUN_TEST_GROUP( Provisioning_Unit_API ); */
```
Comment (gkwicker) [inline libraries/aws/provisioning/test/aws_iot_tests_provisioning.c]: Delete code rather than commenting it out.  Same comment at lines 49/53

File: libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c [hunk 5]
```diff
@@ -141,6 +162,63 @@ TEST( Provisioning_Unit_Serializer, TestSerializeCreateKeysAndCertificatePayload
...
+TEST( Provisioning_Unit_Serializer, TestSerializeCreateCertFromCsrPayloadWithBuffer )
+{
+    uint8_t testBuffer[ sizeof( _expectedSerialization ) ] = { 0 };
+    size_t bufferSize = sizeof( testBuffer );
+
+    /* Test the serializer function. */
+    TEST_ASSERT_EQUAL( true,
+                       _AwsIotProvisioning_SerializeCreateCertFromCsrRequestPayload( _testCsrString,
+                                                                                     strlen( _testCsrString ),
+                                                                                     &testBuffer[ 0 ],
```
Comment (gkwicker) [inline libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c]: You can improve this type of test by reserving a few `uint8_t`s at the beginning and end of `testBuffer`, filling them with some known values (e.g. `0xa5`), and then calling your routine under test with `&testBuffer[ X ]`, where `X` is after your special filled area.  After calling your routine you can then verify that your special values are still in place at the beginning and end of the test array.  This can help catch some common overrun errors.

^ In general I like to capture that in a common test utility that is used throughout all code which does any sort of buffer formatting.

**Review by gkwicker — CHANGES_REQUESTED** [review]
See comments

File: libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c (hunk 2 above)
Comment (aggarw13) [reply libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c]: Yes, this is old code. I plan to create a hygiene epic for making all quality improvements I have documented here: https://quip-amazon.com/56ZDAgTB0Ln5

File: libraries/aws/provisioning/test/aws_iot_tests_provisioning.c (hunk 4 above)
Comment (aggarw13) [reply libraries/aws/provisioning/test/aws_iot_tests_provisioning.c]: Ohh, accidentally pushed this. Will revert the change.

File: libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c (hunk 3 above)
Comment (aggarw13) [reply libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c]: Yes, good suggestion. I will break this function into 2 for better readability!

File: libraries/aws/provisioning/src/aws_iot_provisioning_api.c (hunk 1 above)
Comment (aggarw13) [reply libraries/aws/provisioning/src/aws_iot_provisioning_api.c]: Created a separate function for doing the calculation of serialization payload size for better readability and single-responsibility of the serializer function!

File: libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c (hunk 5 above)
Comment (aggarw13) [reply libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c]: That's a good suggestion. Incorporated it in the test

File: libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c [hunk 6]
```diff
@@ -355,30 +360,168 @@ AwsIotProvisioningError_t _AwsIotProvisioning_SerializeCreateKeysAndCertificateR
...
+    if( isDrySerialization == true )
+    {
+        isSuccessStatus = _checkSuccessOrBufferToSmall;
+    }
+    else
+    {
+        isSuccessStatus = _checkSuccess;
+    }
+
+    if( status == true )
```
Comment (muneebahmed10) [inline libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c]: Why would `status` not be true at this point? It doesn't get changed before this

File: libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c [hunk 7]
```diff
@@ -355,30 +360,168 @@ AwsIotProvisioningError_t _AwsIotProvisioning_SerializeCreateKeysAndCertificateR
...
+    {
+        /* Perform a dry-run serialization of the Certificate-Signing Request */
+        /* data to calculate the size of the payload. */
+        if( _serializeCertFromCsrPayload( pCertificateSigningRequest,
+                                          csrLength,
+                                          &outerEncoder,
+                                          true ) == false )
+        {
+            status = false;
+        }
```
Comment (muneebahmed10) [inline libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c]: If you're not going to log anything, then 
```
status = _serializeCertFromCsrPayload( pCertificateSigningRequest,
                                       csrLength,
                                       &outerEncoder,
                                       true);
```
 would save you a few lines

File: libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c [hunk 8]
```diff
@@ -355,30 +360,168 @@ AwsIotProvisioningError_t _AwsIotProvisioning_SerializeCreateKeysAndCertificateR
...
+    else
+    {
+        /* Serialize the Certificate-Signing Request string in the buffer. */
+        if( _serializeCertFromCsrPayload( pCertificateSigningRequest,
+                                          csrLength,
+                                          &outerEncoder,
+                                          false ) == false )
+        {
+            status = false;
+        }
```
Comment (muneebahmed10) [inline libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c]: Same as above

File: libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c [hunk 9]
```diff
@@ -54,6 +85,7 @@ static const char * _testCertificateToken = "Token!";
 /**
  * @brief A sample list of parameters entries for testing serialization logic.
  */
+
```
Comment (muneebahmed10) [inline libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c]: Doesn't look like this newline was intentional

File: libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c [hunk 10]
```diff
@@ -64,6 +96,59 @@ static const size_t _numOfSampleParameters = 3;
...
+    for( int index = 0; index < _reserveSize; index++ )
+    {
+        assert( _bufferOverrunCheckValue == buffer[ index ] );
+    }
+
+    for( int index = totalBufferSize - 1; index >= totalBufferSize - _reserveSize; index-- )
+    {
+        assert( _bufferOverrunCheckValue == buffer[ index ] );
+    }
```
Comment (muneebahmed10) [inline libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c]: Looks like these two `for`s could be combined into one with `buffer[ totalBufferSize - 1 - index ]`. Just a style choice though, and it doesn't really matter since this is only a unit test file.

File: libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c [hunk 11]
```diff
@@ -141,6 +228,71 @@ TEST( Provisioning_Unit_Serializer, TestSerializeCreateKeysAndCertificatePayload
...
+TEST( Provisioning_Unit_Serializer, TestSerializeCreateCertFromCsrPayloadWithBuffer )
+{
+    size_t reserveOffset;
+
+    /* Allocate buffer that will be used for serialization.
+     * Note: We will allocate more space than the expected serialization size to perform
+     * buffer overrun checks after the serialization operation.  */
+    size_t serializationSize = sizeof( _expectedSerialization );
+    uint8_t * testBuffer = _allocateBufferMemoryWithBufferOverrunReserve(
+        serializationSize, &reserveOffset );
```
Comment (muneebahmed10) [inline libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c]: Is this spacing a result of uncrustify? I thought when we split functions we only have one parameter per line

File: libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c [hunk 12]
```diff
@@ -141,6 +228,71 @@ TEST( Provisioning_Unit_Serializer, TestSerializeCreateKeysAndCertificatePayload
...
+
+    /* Test the serializer function. */
+    TEST_ASSERT_EQUAL( true,
+                       _AwsIotProvisioning_SerializeCreateCertFromCsrRequestPayload( _testCsrString,
+                                                                                     strlen( _testCsrString ),
+                                                                                     testBuffer + reserveOffset,
+                                                                                     serializationSize ) );
+    /* Verify the generated serialization in the buffer. */
+    TEST_ASSERT_EQUAL( 0, memcmp( _expectedSerialization, testBuffer + 5,
```
Comment (muneebahmed10) [inline libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c]: Same as above

File: libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c (hunk 11 above)
Comment (aggarw13) [reply libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c]: Nope, it was manual. I will keep update to parameters on different lines.

Comment (aggarw13) [reply libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c]: Ideally, uncrustify should have automatically separated the parameters on new lines when formatting this code text.

File: libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c (hunk 12 above)
Comment (aggarw13) [reply libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c]: Changed

File: libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c (hunk 10 above)
Comment (aggarw13) [reply libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c]: Had done for better readability. I suppose that having in the same loop would be readable as well with comments. Made the change

File: libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c (hunk 9 above)
Comment (aggarw13) [reply libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c]: Removed

File: libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c (hunk 8 above)
Comment (aggarw13) [reply libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c]: Updated

File: libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c (hunk 7 above)
Comment (aggarw13) [reply libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c]: Good observation.

File: libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c (hunk 6 above)
Comment (aggarw13) [reply libraries/aws/provisioning/src/aws_iot_provisioning_serializer.c]: Yes, was a copy-paste carry-over from a previous version in this branch that didn't have the common function (but only a single serializer function).
Have removed it.

File: libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c [hunk 13]
```diff
@@ -64,6 +95,57 @@ static const size_t _numOfSampleParameters = 3;
...
+ * is not corrupted by buffer overrun errors in the serializer function under
+ * test.
+ */
+void _checkForBufferOverrun( const uint8_t * buffer,
+                             size_t sizeOfSerialization )
+{
+    size_t totalBufferSize = sizeOfSerialization +
+                             _reserveSize + _reserveSize;
+
+    /* Check that both the front and read reserve spaces in the buffer are not */
```
Comment (muneebahmed10) [inline libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c]: `read` -> `rear`

File: libraries/aws/provisioning/test/aws_iot_tests_provisioning.c (hunk 4 above)
Comment (muneebahmed10) [reply libraries/aws/provisioning/test/aws_iot_tests_provisioning.c]: Will this be reverted as part of this PR?

Comment (gkwicker) [reply libraries/aws/provisioning/test/aws_iot_tests_provisioning.c]: Looks like you still need to remove this commented-out code

**Review by gkwicker — APPROVED** [review]

File: libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c (hunk 13 above)
Comment (aggarw13) [reply libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c]: Fixed

**Review by muneebahmed10 — APPROVED** [review]
